vllm/worker/worker.py

```diff
@@ -233,43 +239,68 @@ def _prepare_inputs(
             seq_groups.append((seq_ids, sampling_params))
 
             num_seqs = len(seq_ids)
+            seq = seq_group_metadata.seq_data[seq_ids[0]]
+            if len(seq.draft_token_probs) > 0:
+                assert num_seqs == 1
+                selected_token_end_idx = selected_token_start_idx + len(seq.draft_token_probs) + 1
+            else:
+                selected_token_end_idx = selected_token_start_idx + num_seqs
             selected_token_indices.extend(
                 range(selected_token_start_idx,
-                      selected_token_start_idx + num_seqs))
-            selected_token_start_idx += num_seqs
+                      selected_token_end_idx))
+            selected_token_start_idx = selected_token_end_idx
 
             categorized_sample_indices[sampling_params.sampling_type].extend(
                 range(categorized_sample_indices_start_idx,
                       categorized_sample_indices_start_idx + num_seqs))
             categorized_sample_indices_start_idx += num_seqs
 
             for seq_id in seq_ids:
-                seq_data = seq_group_metadata.seq_data[seq_id]
-                generation_token = seq_data.get_last_token_id()
-                input_tokens.append([generation_token])
-
-                context_len = seq_data.get_len()
-                position = context_len - 1
-                if self.sliding_window is not None:
-                    context_len = min(context_len, self.sliding_window)
-                input_positions.append([position])
-
+                seq_data: SequenceData = seq_group_metadata.seq_data[seq_id]
+                
                 block_table = seq_group_metadata.block_tables[seq_id]
-
+                draft_len = len(seq_data.get_draft_token_ids())
+                prompt_len = seq_data.get_len()
+                context_len = prompt_len - draft_len
                 max_context_len = max(max_context_len, context_len)
                 max_num_blocks_per_seq = max(max_num_blocks_per_seq,
-                                             len(block_table))
+                                                len(block_table))
                 context_lens.append(context_len)
-
-                block_number = block_table[position // self.block_size]
-                block_offset = position % self.block_size
-                slot = block_number * self.block_size + block_offset
-                slot_mapping.append([slot])
-
-                if self.sliding_window is not None:
-                    sliding_window_blocks = (self.sliding_window //
-                                             self.block_size)
-                    block_table = block_table[-sliding_window_blocks:]
+                if draft_len > 0:
```

Comment:
I think here you'd better add some comments like:
"Speculative Decoding enabled case: <explain the main difference of this case compared with normal case, maybe something like it need to pass in multiple tokens that are accepted in previous run>"